client.go

```diff
@@ -39,32 +46,39 @@ type Client struct {
 	jobLock     sync.RWMutex
 }
 
-type partialJob struct {
-	data, warnings io.WriteCloser
-}
-
 // Close terminates the connection to the server
 func (c *Client) Close() error {
-	c.conn.Close()
 	// TODO: figure out when to close packet chan
-	return nil
+	return c.conn.Close()
 }
 
 func (c *Client) submit(fn string, payload []byte, data, warnings io.WriteCloser, t packet.Type) (*job.Job, error) {
+	// create and marshal the gearman packet
 	pack := &packet.Packet{
 		Code:      packet.Req,
 		Type:      t,
 		Arguments: [][]byte{[]byte(fn), []byte{}, payload},
 	}
-	b, err := pack.MarshalBinary()
+	buf, err := pack.MarshalBinary()
 	if err != nil {
 		return nil, err
 	}
-	if _, err := io.Copy(c.conn, bytes.NewBuffer(b)); err != nil {
+
+	// write the packet to the gearman server
+	if _, err := io.Copy(c.conn, bytes.NewBuffer(buf)); err != nil {
 		return nil, err
 	}
+
+	// block while the client waits for confirmation that a job has been created
 	c.partialJobs <- &partialJob{data: data, warnings: warnings}
-	return <-c.newJobs, nil
+
+	// wait for confirmation, potentially times out
+	select {
+	case j := <-c.newJobs:
+		return j, nil
+	case <-time.After(10 * time.Second):
```

Comment:
agreed, I'm not convinced that this is the right move, but it seems really odd that the client is entirely blocked while waiting for a response.